### PR TITLE
Add blobs to `eth_feeHistory`

### LINKED ISSUE #6580: Add blobs to `eth_feeHistory`
Blob gas price oracle and historical API

see https://github.com/ethereum/execution-apis/pull/486

### PR DESCRIPTION
**Pull request by Gabriel-Trintinalia**
## PR description
Add blob fields to eth_fee_History

## Fixed Issue(s)
fixes #6580

On Sepolia:
`curl -H "Content-Type: application/json" -d '{"id": 1, "jsonrpc": "2.0", "method": "eth_feeHistory", "params": ["0x5", "0x526006", [20,30]] }' http://127.0.0.1:8545/ | jq .
`
```
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "oldestBlock": "0x526002",
    "baseFeePerGas": [
      "0x4506f",
      "0x43d32",
      "0x44d30",
      "0x42021",
      "0x39c1d",
      "0x38cd0"
    ],
    "baseFeePerBlobGas": [
      "0x4a1bd3218",
      "0x502994693",
      "0x4d137e31b",
      "0x4741676a3",
      "0x3f5694c1f",
      "0x4741676a3"
    ],
    "gasUsedRatio": [
      0.4303144,
      0.5589600333333333,
      0.3363257666666667,
      0,
      0.4337706333333333
    ],
    "blobGasUsedRatio": [
      0.8333333333333334,
      0.3333333333333333,
      0.16666666666666666,
      0,
      1
    ],
    "reward": [
      [
        "0x186a1",
        "0x186a1"
      ],
      [
        "0x186a1",
        "0x186a1"
      ],
      [
        "0x59682f00",
        "0x59682f00"
      ],
      [
        "0x0",
        "0x0"
      ],
      [
        "0x3b9aca00",
        "0x3b9aca00"
      ]
    ]
  }
}
```

### CODE REVIEW

**Comment by siladu** [comment]
> Zeroes are returned for pre-EIP-4844 blocks.

Has this been tested on a real node? I guess mainnet nodes should return all 0s?

**Comment by Gabriel-Trintinalia** [comment]
> > Zeroes are returned for pre-EIP-4844 blocks.
> 
> Has this been tested on a real node? I guess mainnet nodes should return all 0s?

@siladu  Sepolia^ yeah, I pre-cancun node must return all zeros

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java [hunk 1]
```diff
@@ -332,25 +345,93 @@ private List<Wei> getBaseFees(final List<BlockHeader> blockHeaders) {
...
+  private List<BlockHeader> prependParent(final List<BlockHeader> blockHeaders) {
+    List<BlockHeader> extendedBlockHeaders = new ArrayList<>();
+    Optional<BlockHeader> parent = blockchain.getBlockHeader(blockHeaders.get(0).getParentHash());
+    parent.ifPresent(extendedBlockHeaders::add);
+    extendedBlockHeaders.addAll(blockHeaders);
+    return extendedBlockHeaders;
+  }
+
+  private Wei getBlobGasFee(final ProtocolSpec spec, final BlockHeader parent) {
+    return spec.getFeeMarket().blobGasPricePerGas(calculateExcessBlobGasForParent(spec, parent));
```
Comment (siladu) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java:395]: We don't cache this on the BlockHeader like baseFee?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java [hunk 2]
```diff
@@ -332,25 +345,93 @@ private List<Wei> getBaseFees(final List<BlockHeader> blockHeaders) {
...
   }
 
+  private List<Wei> getBlobBaseFees(final List<BlockHeader> blockHeaders) {
+    if (blockHeaders.isEmpty()) {
+      return Collections.emptyList();
+    }
+    // Extend block headers with their first parent
+    List<BlockHeader> extendedBlockHeaders = prependParent(blockHeaders);
+
+    // Calculate base fees for each pair of parent and headers.
```
Comment (siladu) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java:355]: Why do we need to recalculate?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java [hunk 3]
```diff
@@ -332,25 +345,93 @@ private List<Wei> getBaseFees(final List<BlockHeader> blockHeaders) {
...
   }
 
+  private List<Double> getBlobGasUsedRatios(final List<BlockHeader> blockHeaders) {
+    return blockHeaders.stream().map(this::calculateBlobGasUsedRatio).toList();
+  }
+
+  private double calculateBlobGasUsedRatio(final BlockHeader blockHeader) {
+    ProtocolSpec spec = protocolSchedule.getByBlockHeader(blockHeader);
+    long blobGasUsed = blockHeader.getBlobGasUsed().orElse(0L);
+    double currentBlobGasLimit = spec.getGasLimitCalculator().currentBlobGasLimit();
```
[the remark this answers is not in the document]
Comment (Gabriel-Trintinalia) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java:399]: Yes, it is

File: ethereum/api/src/test/resources/org/hyperledger/besu/ethereum/api/jsonrpc/eth/eth_feeHistory_fieldNamesWithReward.json [hunk 4]
```diff
@@ -23,10 +23,12 @@
         "0x3437004a",
         "0x2dbc88c1"
       ],
+      "baseFeePerBlobGas" : [ "0x0", "0x1", "0x1" ],
       "gasUsedRatio": [
         0.004079142040086682,
         0.003713085594819442
       ],
+      "blobGasUsedRatio" : [ 0.0, 0.3333333333333333 ],
```
[the remark this answers is not in the document]
Comment (Gabriel-Trintinalia) [reply ethereum/api/src/test/resources/org/hyperledger/besu/ethereum/api/jsonrpc/eth/eth_feeHistory_fieldNamesWithReward.json:31]: So for baseFeePerBlobGas the method returns all blob fees for the requested range + the next one. for the ratio, we return only for the requested range.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java [hunk 5]
```diff
@@ -332,25 +345,93 @@ private List<Wei> getBaseFees(final List<BlockHeader> blockHeaders) {
         .toList();
   }
 
+  private List<Wei> getBlobBaseFees(final List<BlockHeader> blockHeaders) {
+    if (blockHeaders.isEmpty()) {
+      return Collections.emptyList();
+    }
+    // Extend block headers with their first parent
+    List<BlockHeader> extendedBlockHeaders = prependParent(blockHeaders);
```
[the remark this answers is not in the document]
Comment (Gabriel-Trintinalia) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java]: I have removed that comment. The nextBaseFee is calculated above in the class

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java (hunk 2 above)
Comment (Gabriel-Trintinalia) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java]: We do not save the blobBase fee in the header

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java (hunk 1 above)
Comment (Gabriel-Trintinalia) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java]: nop, blobBaseFee is calculated using the parent header

**Review by macfarla — COMMENTED** [review]
needs a changelog entry

**Review by macfarla — APPROVED** [review]
LGTM

**Pull request merged**